### ISSUE TITLE
RSA KeyGen bitLens values don't match outputted bit lengths.

### ISSUE THREAD

**Issue by mniforatos** [usnistgov/ACVP-Server#67]
**Environment**
Both Production and Demo

**testSessionId (with vsID in parenthesis)**
Prod: 4428 (vsID: 42161)
Demo: 138194 (vsID: 433397)

**Algorithm registration**
```
[
	{
		"acvVersion":"1.0"
	},
	{
	  "algorithms": [
		{
		  "mode": "keyGen",
		  "capabilities": [
			{
			  "randPQ": "B.3.6",
			  "properties": [
				{
				  "modulo": 2048,
				  "primeTest": [
					"tblC3"
				  ]
				},
				{
				  "modulo": 3072,
				  "primeTest": [
					"tblC3"
				  ]
				}
			  ]
			}
		  ],
		  "pubExpMode": "random",
		  "keyFormat": "standard",
		  "algorithm": "RSA",
		  "revision": "FIPS186-4"
		}
	  ]
	}
]
```

**Endpoint in which the error is experienced**
https://acvts.nist.gov/acvp/v1/testSessions
https://demo.acvts.nist.gov/acvp/v1/testSessions

**Expected behavior**
For vsID: 433397, tcId 1, the bitLens from the request file are 577, 157, 388, and 265. However, the corresponding xP1, xP2, xQ1, and xQ2 are 584, 160, 392, and 272, respectively. At first I thought this was just due to the ACVP server getting rid of the leading 0's. For example, xQ1 is: 0A46FBA94A492B77203B0C61E91C9C1170C48809F03558A8C15AC45E362211E969F17F7059DED181B96BD820DF2CEA162F. There are 4 leading 0's due to the '0' at the beginning. The length of xQ1 is 4 * 98 hex characters, or 392 bits, but the corresponding value of bitLen is only 388. The value of 388 makes sense if you remove the 4 leading 0's.

However, the discrepancy I've noticed doesn't always directly correlate to the amount of leading 0's. For example, xP1 is: 00E58ECFEC80D9A0EA8170808F6C89F07E4618655923BE8A68C471861E79D6FA4ED996CA87ECB6CF2D560B283A582A8DBB8C6E77686FDD83AF15BCA35006F7F6AAABD8B9FA447CECBD. This equates to 4 * 146 hex characters = 584 bits. The '00' at the beginning is '0000 0000' in binary. If you remove the 8 leading 0's, then xP1 is 576 bits long. However, the corresponding value from bitLen is 577 instead of 576. 

Similarly, in this example xP2 is: 01976E70B7C7FDE86AAFEF651C346EF5A06D2103, or 4 * 40 hex characters = 160 bits. The '01' at the beginning is 0000 0001 in binary, so there's 7 leading 0's. However, instead of bitLen listing 153 bits for xP2, it lists 157.

It appears that the bitLen seems to be getting miscalculated server-side, but I'm not really sure what exactly is going on.

Thanks!

**Comment by celic** [comment]
I think the server is correct here. The server will always communicate bitstrings as complete bytes, as it is in this case. Each of the values is left-padded to the next complete byte. If you take the rightmost `bitlen` bits of each value, you should end up with the correct result.

**Comment by mniforatos** [comment]
Hello @celic, thanks for the reply!

Yes, the server is indeed communicating the bitstrings as complete bytes. However, as evidenced in my examples above, the values provided for bitLens don't match up with the rightmost bits as you said. For instance, refer to my second example, xP2. If we use the ‘bitLens’ value the server provides (157) we end up with effectively 4 leading zeroes remaining, which shows that it has been calculated wrong server-side as you shouldn’t ever end up with leading zeros, which I believe is what you were saying.

Similarly with my first example, xP1. The character length is 584 bits, but taking the rightmost bits of each value you end up with 576 bits, due to there being 8 leading 0s. However, the server calculated this as 577 bits instead of 576, so there's still one leading zero remaining.

Thanks!

**Comment by EH-Acumen** [comment]
Hello @celic.  I am interested in this issue as well.  I agree with @mniforatos.  I do not see how the bitlen provided by the server is the correct result.

**Comment by celic** [comment]
I think the number of leading 0s isn't relevant. Take the rightmost bits as is. The first bits in the value are allowed to be 0. The argument could be made that a 512-bit number with a leading 0-bit could also be a 511-bit number. In this case as these are random values, we don't modify or force the resulting values to have a leading 1-bit. FIPS186-4 isn't definitive on this requirement. 

Am I understanding the issue correctly? You are assuming that the first bit in the value bitlens describes must be a 1?

**Comment by EH-Acumen** [comment]
@celic perhaps the question is more of how are the random values xP1, xP2, xQ1, and xQ2 generated from the bitlen given?  The length of the value is given in binary but the value is provided in hexadecimal.  It is unsure how from the example above the xP1 value with given bitlen 577 is 00E58ECFEC80D9A0EA8170808F6C89F07E4618655923BE8A68C471861E79D6FA4ED996CA87ECB6CF2D560B283A582A8DBB8C6E77686FDD83AF15BCA35006F7F6AAABD8B9FA447CECBD. This hexadecimal value appears to be 576 bits long, ignoring the leading zeros, and not 577 bits long. Is it the case that whatever value is provided for xP1, XP2, xQ1 and xQ2 only the rightmost number of bits from the bitlen value provided should be used to determine the values p and q?

**Comment by celic** [comment]
We generate the values by getting 577 random bits. The leftmost bits of that value _can_ be 0. There are no restrictions. Then we need to communicate the incomplete bytes as completed bytes with hex, so 0s are prepended to the value. This may lead to a full byte of 0s upfront. This is not a bug, it is just communicating the 577-bit value. Take the rightmost 577 bits of the full hex value. Don't drop the leading 00 byte.

**Comment by mniforatos** [comment]
Hello @celic ,

So are you saying that since the values generated can include a 0 in the leftmost bit, that the server will relay that information in the form of bitLens? And then, since the server still only counts the rightmost bits, that the "leading 0s" are still dropped, which means in this example 7 leading 0s are dropped and the 8th 0 is still counted? Cause if you don't drop the 7 leading 0's, bitLens should be 584 not 577.

In the example I gave for xP2, does that mean the server generated a 0 and 1 for the leftmost bits, which equates to 0000 0001?Then, taking the rightmost bits you drop the first 3 0's and keep the 4th one since the 4th one was an actual generated value. If so, then the value of 157 provided in bitLens makes sense since the character value is 160, but we're dropping the first 3 0s and keeping the 4th (i.e. only taking the rightmost bits).

Does that sound correct?

Thanks!

Best Regards,
Matt Niforatos

**Comment by celic** [comment]
Your first statement is correct. Take the 577 rightmost bits from the full 584-bit value. So 7 0bits are dropped off the front, and one 0 bit from that leading byte is used within the value. For the xP2 example, you would ignore the first three bits (which MUST be 0s, this is something you could verify potentially) and take the remaining bits whatever they may be.

**Comment by mniforatos** [comment]
Hello @celic ,

Okay, thank you for the confirmation. So it seems that the only way to know if a 0 was generated for the leftmost bit is to look at the value provided in bitLens instead of by looking at/doing something with the values for xP1, xP2, xQ1, or xQ2. I say that because for xP2, the leftmost bit was clearly a 0 due to bitLens being 157, but for xQ1 the character length was 392 and the corresponding bitLens value was 388, which means a 0 was not the leftmost bit.

In other words, if you just look at xP2 and xQ1, which is what some vendors' test harnesses do, you'll see 01 and 0A at the beginning of each, respectively. Without referring to the values provided for each in bitLens, there's no way to know if the 0 at the beginning of each includes a generated 0 or if it's just padded 0s.

I just want to confirm this is the case before I go back to our vendors and explain it to them, so that they can update their test harnesses properly.

Thanks!

Best Regards,
Matt Niforatos

**Comment by celic** [comment]
Yes, the bitlens MUST be used to construct the value properly in all cases. I believe this was the case in the original testing from CAVS as well. If you have any recommendations for clarifying this in the specification, we can make those changes too.

**Comment by mniforatos** [comment]
Hello @celic ,

So, I did a little more testing and found that the pattern discussed above does not actually always match the output. Take for example Demo Server Test Session 169871, vsID: 517238. For tcId: 1, bitlens is 371, 196, 268, 338. The corresponding value for xP1 is:

001031CA6DA66BB3F5C2E116ED199B739B1FE2D34EDD117BF8951E4D3B9293806AFD043198E9C3E6439DF00D98D5E5

That’s 94 characters or 376 bits, and the bitlens value for it is 371. However, if we take the rightmost bits as described above, you’d get either 373 (if the first two generated values were both 0), 369 (if only the first generated value was 0 and then there was an extra padded 0), or 365 (if the first generated value was 1 and there were 2 padded 0s). Since none of those numbers is 371, my assumption is that the server calculated bitlens wrong.

Another example is tcId 2, where bitlens is 443, 291, 386, 246. The corresponding value for xP1 is:

01589D8BBE6593EF8E2AB70DDE38291ADDAEADBA5F5544F516B1B5A4AE098FCB878251A717EE68AEC417F859B78E769EAE17656AD83F9A23.

That's 112 characters or 448 bits, and the bitlens value for it is 443. However, if we take the rightmost bits as described above, you'd get either 445 (if the first generated value was 0) or 441 (if the first generated value was 1 with one padded 0). Again, since neither of these numbers is 443 it would seem that the server calculated bitlens wrong.

What accounts for this discrepancy, since it doesn't match the pattern/method you described above?

Thanks!

Best Regards,
Matt

**Comment by celic** [comment]
There can be up to 7 padded 0 bits from the front. The value itself can
have any number of leading 0s. There are no restrictions from the server.
Take the full rightmost bits based on the bitlens.

**Comment by mniforatos** [comment]
Thanks for the response, but can you please address my comment about the discrepancy above? Based on the actual value for xP1, and based on your description above about the possibility that a 0 may be the leftmost generated value(s), the only possible values for the bit length of xP1 are 373, 369, or 365, as I described above. The value of 371 can't be the bit length value of xP1. You mention that the server doesn't have any restrictions on the number of leading/padded 0s (up to 7), however leading/padded 0s should not be counted in the bit length.

Based on the actual provided value of xP1, how does the server calculate a bit length of 371?

**Comment by celic** [comment]
The server creates a bitlen then generates a random value for it, with no
restrictions on the value. The value could theoretically be full 0 bits.
Still the user should grab the rightmost bitlen bits and use them
regardless of the value. You are not to force a leading 1 in the value.

This is how CAVS would generate values when RSA testing was originally
introduced, and has not been changed for ACVP.

**Comment by powersmc** [comment]
Just to add a bit to this discussion, I think the primary issue here is just that CAVS used the more "standard" definition of [bit length](https://mathworld.wolfram.com/BitLength.html) of a number, whereas the ACVP seems to basically be defining bit length as "the length of the buffer, in bits, that we used to generate this random number".  Logically I think since FIPS 186-4 is written around numbers (xp1, xp2, xq1, xq2) and not buffers it makes the most sense to use CAVS' definition.

I confirmed by generating request file in CAVS that this is the case - CAVS never includes the leading 0s in the calculation of bit length (see attached files for my quick-and-dirty script, the request file from CAVS, and the output).

In short - we have vendors that have found that they get the **correct** answers if they ignore the bitlen values provided by the ACVP server and instead calculate them themselves based on the integer values of xp1, xp2, xq1, xq2 (such as by using python's built in '[bit_length](https://python-reference.readthedocs.io/en/latest/docs/ints/bit_length.html)' function), and that they're getting the wrong answers if they use the bitlen values from the server.

[CAVS_bitlen_test.zip](https://github.com/usnistgov/ACVP-Server/files/6243365/CAVS_bitlen_test.zip)

**Comment by celic** [comment]
I see the opposite, and this was used within our development and internal testing. I have CAVS vector sets that do the same as what ACVP does in these cases. Here is an example: 

```
bitlen4 = 210
Xq2 = 0053fe0077776e844771ce4c763bda972ba454e205b7ab91951fcb
```

**Comment by powersmc** [comment]
That seems very inconsistent with the vectors / script I attached to the previous post.  That was generated with CAVS 21.4 this morning, and there was not a single instance in the 200 vectors produced where bitlen included leading zeroes.

I agree that with your example above that the bitlen of 210 does include three leading 0-bits as it should be 207.  It's unclear, however, what the explanation for that is.  Which CAVS version and options were used to generate those vectors that you're referencing?  I replicated the options from the [original post](https://github.com/usnistgov/ACVP-Server/issues/67#issue-789252177) when I generated mine.

**Comment by celic** [comment]
I don't recall the version. The file we use for testing was generated 3 years ago at least. It may have even been pulled from CSRC and may date back to the original publication of 186-3 testing. I need to double check this.

**Comment by celic** [comment]
This behavior occurs consistently in the KeyGen_186-3.rsp available here https://csrc.nist.gov/CSRC/media/Projects/Cryptographic-Algorithm-Validation-Program/documents/dss/186-3rsatestvectors.zip

This was generated using CAVS 11.1 back in 2011. I'm now trying to check changelogs on CAVS to see if this was behavior that was intentionally updated in later versions.

**Comment by celic** [comment]
This was changed in CAVS 11.3, which forced the leading bit to be a 1. The vector sets on CSRC were never updated. That means the behavior is currently incorrect in ACVTS. 

I vaguely remember toying with this while implementing the RSA algorithms but got incorrect results with our previous tests because the CAVS file was from version 11.1. At the time, we didn't have a way of scanning changelogs.

**Comment by powersmc** [comment]
Ah interesting.  Thanks for looking into this!

**Comment by celic** [comment]
Thank you for the discussion and detail here. I'll try to get this into an update soon. Looks like CSRC needs an update in addition.

**Comment by Kritner** [comment]
Hey @mniforatos, @powersmc @EH-Acumen I'm taking a look into this, can you share if you've seen this sort of incorrect behavior in any of the other prime generation techniques aside from "Generation of Probable Primes with Conditions Based on Auxiliary
Probable Primes"?  I believe I'm following the issue, but if you've seen it elsewhere that would save me some digging.

**Comment by powersmc** [comment]
Hey @Kritner - I am not aware of this behavior being present in other prime generation techniques.  B.3.6 is the only one I've personally observed it in.

**Comment by mniforatos** [comment]
Hello @Kritner - B.3.6 is the only one I've observed it in as well, but I haven't tested any of the other options.

**Comment by Kritner** [comment]
This change is on demo in release [v1.1.0.17](https://github.com/usnistgov/ACVP-Server/releases/tag/v1.1.0.17)

**Comment by Kritner** [comment]
This change is on prod as of release [v1.1.0.17](https://github.com/usnistgov/ACVP-Server/releases/tag/v1.1.0.17)
